xdsl/dialects/bufferization.py

```diff
@@ -42,42 +47,29 @@ class TensorMemrefInferenceConstraint(VarConstraint[Attribute]):
     and checks for matching element type, shape (ranked only), as well as verifying sub constraints.
     """
 
-    def infer(self, constraint_context: ConstraintContext) -> Attribute:
-        if self.name in constraint_context.variables:
-            m_type = constraint_context.get_variable(self.name)
-            if isa(m_type, MemRefType[Attribute]):
-                return TensorType(m_type.get_element_type(), m_type.get_shape())
-            if isa(m_type, UnrankedMemrefType[Attribute]):
-                return UnrankedTensorType(m_type.element_type)
-        raise ValueError(f"Unexpected {self.name} - cannot infer attribute")
+    memref_var_constr: VarConstraint[
+        MemRefType[Attribute] | UnrankedMemrefType[Attribute]
+    ]
+
+    def can_infer(self, variables: Set[str]) -> bool:
+        return self.memref_var_constr.can_infer(variables)
+
+    def infer(
+        self, variables: dict[str, ResolveType]
+    ) -> TensorType[Attribute] | UnrankedTensorType[Attribute]:
+        memref_type = self.memref_var_constr.infer(variables)
+        if isinstance(memref_type, MemRefType):
+            return TensorType(memref_type.element_type, memref_type.shape)
+        return UnrankedTensorType(memref_type.element_type)
 
     def verify(self, attr: Attribute, constraint_context: ConstraintContext) -> None:
-        if self.name in constraint_context.variables:
-            seen = constraint_context.get_variable(self.name)
-            if not (
-                isinstance(attr, ContainerType)
-                and isinstance(seen, ContainerType)
-                and attr.get_element_type() == seen.get_element_type()
-            ):
-                raise VerifyException(
-                    f"Unexpected {self.name} - cannot verify element type of attribute {attr}"
-                )
-            if (
-                isinstance(attr, ShapedType) != isinstance(seen, ShapedType)
-                or isinstance(attr, ShapedType)
-                and isinstance(seen, ShapedType)
-                and attr.get_shape() != seen.get_shape()
-            ):
-                raise VerifyException(
-                    f"Unexpected {self.name} - cannot verify shape of attribute {attr}"
-                )
-        elif isinstance(attr, ContainerType):
-            self.constraint.verify(attr, constraint_context)
-            constraint_context.set_variable(self.name, attr)
-        else:
-            raise VerifyException(
-                f"Unexpected {self.name} - attribute must be ContainerType"
-            )
+        if isa(attr, TensorType[Attribute]):
+            memref_type = MemRefType(attr.element_type, attr.shape)
+            return self.memref_var_constr.verify(memref_type, constraint_context)
+        if isa(attr, UnrankedTensorType[Attribute]):
+            memref_type = UnrankedMemrefType.from_type(attr.element_type)
+
+        raise VerifyException(f"Expected TensorType or UnrankedTensorType, got {attr}")
```

Comment:
the logic here seems to be changed

Reply:
It is, though it appears I forgot to amend the doc comment. In my opinion the previous constraint was an abuse of the particular workings of the previous system. The one I have written here is a bit more well-behaved with respect to inference, and can be used to achieve the same effect as the old one. I'll update the doc comment.

Reply:
Would it make sense to merge this change as a separate PR, while we discuss the details of this one?

Reply:
Sure, I'll get that ready first thing tomorrow